src/utils/internalLinks.js

```diff
@@ -112,18 +64,30 @@ export const decodeHashComponent = (string: string): string => {
 };
 
 /**
- * Parse the operand of a `stream` operator.
+ * Parse the operand of a `stream` operator, returning a stream ID.
  *
- * Return null if the operand doesn't match any known stream.
+ * The ID might point to a stream that's hidden from our user (perhaps
+ * doesn't exist). If so, most likely the user doesn't have permission to
+ * see the stream's existence -- like with a guest user for any stream
+ * they're not in, or any non-admin with a private stream they're not in.
+ * Could be that whoever wrote the link just made something up.
+ *
+ * Returns null if the operand has an unexpected shape, or has the old shape
+ * (stream name but no ID) and we don't know of a stream by the given name.
  */
-const parseStreamOperand = (operand, streamsById, streamsByName): null | Stream => {
+// Why does this parser need stream data? Because the operand formats
+// ("new" and "old") collide, and in choosing which format to apply, we
+// want to know if one or the other would give a real stream we know
+// about. Also, we can't get a stream ID from an "old"-format operand
+// without a mapping from stream names to IDs.
+const parseStreamOperand = (operand, streamsById, streamsByName): null | number => {
   // "New" (2018) format: ${stream_id}-${stream_name} .
   const match = /^([\d]+)(?:-.*)?$/.exec(operand);
-  if (match) {
-    const streamId = parseInt(match[1], 10);
-    const stream = streamsById.get(streamId);
+  const newFormatStreamId = match ? parseInt(match[1], 10) : null;
+  if (newFormatStreamId != null) {
+    const stream = streamsById.get(newFormatStreamId);
     if (stream) {
-      return stream;
+      return stream.stream_id;
```

Comment:
By unconditional, do you mean just conditional on `newFormatStreamId` being non-null? Like:

```js
  if (newFormatStreamId != null) {
    return newFormatStreamId;
  }
```

With that, the first test you quoted would fail on any of these three "expect"s:

```js
        expectStream('311', [numbers], numbers);
        expectStream('311-', [numbersHyphen], numbersHyphen);
        expectStream('311-help', [numbersPlus], numbersPlus);
```

The function would actually give 311 as the stream ID, but the expected stream ID is the one for the stream with the (odd) name "311", or "311-", or "311-help".

Reply:
Hmm, yeah, I guess those tests do already cover this.